### DISCUSSION TITLE
PR review not add comments every time

### DISCUSSION THREAD

**Discussion by Ahmet122** [mlarhrouch/azure-pipeline-gpt-pr-review#23] [General]
Hello, I have been using your PR review application for a few days. It added comments for the first PR, but did not add a comment for subsequent PR's. Can you help with this issue?

**Comment by Ahmet122** [comment]
Hello, I would be very happy if you answer my question.

**Comment by mlarhrouch** [comment]
@Ahmet122 can you describe the issue with more informations ?  what do you mean by "subsequent PR's" ?

**Reply by Ahmet122** [reply]
It works fine for a pipeline the first time, but the same pipeline leaves no comments for the next pull request.

**Reply by mlarhrouch** [reply]
can you provide the logs from the build pipeline job for the PRs which no comments are added ?

**Reply by Ahmet122** [reply]
Review of src/ArGPT.Business/Applications/Configuration/KnowledgeBaseFileOptions.cs completed.
New comment added.
Review of src/ArGPT.Business/Applications/Abstract/IApplicationKnowledgeBaseService.cs completed.
New comment added.
Review of src/ArGPT.Business/Applications/Specification/KnowledgeBaseDocumentAppIdSpecification.cs completed.
Review of scripts/createMigrationFile.ps1 completed.
New comment added.
Review of src/ArGPT.Core/Infrastructure/Exception/InfrastructureExceptionBase.cs completed.
Review of src/ArGPT.Core/Infrastructure/Persistence/IRepository.cs completed.
New comment added.
Review of src/ArGPT.Core/Util/Linq/LinqExtensions.cs completed.
New comment added.
Review of src/ArGPT.DatabaseMigrator/Migrations/0017-add-Permission-Role-RolePermission-.sql completed.
Review of src/ArGPT.Dtos/Applications/ApplicationKnowledgeBaseDocumentResponse.cs completed.
Review of src/ArGPT.Core/Infrastructure/FileAccess/IFileStore.cs completed.
New comment added.
Review of src/ArGPT.Core/Infrastructure/Persistence/EfRepositoryBase.cs completed.
Review of src/ArGPT.Entities/Applications/ApplicationSettings.cs completed.
New comment added.
Review of src/ArGPT.Entities/Applications/ApplicationKnowledgeBaseDocument.cs completed.
Review of src/ArGPT.Infrastructure/Persistence/Configuration/Applications/ApplicationKnowledgeBaseDocumentEntityTypeConfiguration.cs completed.
Review of src/ArGPT.Infrastructure/Persistence/DbContexts/ArGPTContext.cs completed.
New comment added.
Review of src/ArGPT.Infrastructure/ArGPT.Infrastructure.csproj completed.
Review of src/ArGPT.Entities/Applications/Application.cs completed.
New comment added.
Review of src/ArGPT.DatabaseMigrator/Migrations/20230809-151348-application-knowledge-base.sql completed.
Review of src/ArGPT.Infrastructure/VectorStore/AzureCognitiveSearch/Configuration/AzureCognitiveSearchOptions.cs completed.
New comment added.
Review of src/ArGPT.Infrastructure/Persistence/Repositories/Applications/IApplicationKnowledgeBaseDocumentRepository.cs completed.
Review of src/ArGPT.Infrastructure/VectorStore/VectorStoreRecord.cs completed.
Review of src/ArGPT.WebApi/ArGPT.WebApi.csproj completed.
New comment added.
Review of src/ArGPT.Infrastructure/FileAccess/FileUtils.cs completed.
Review of src/ArGPT.WebApi/appsettings.json completed.
Review of src/ArGPT.WebApi/Filters/ExceptionFilter.cs completed.
New comment added.
Review of src/ArGPT.Business/Applications/Abstract/IApplicationService.cs completed.
Review of src/ArGPT.Infrastructure/VectorStore/IVectorStore.cs completed.
Review of src/Tests/ArGPT.UnitTests/Applications/ApplicationsManagerTest.cs completed.
New comment added.

**Reply by Ahmet122** [reply]
It says new comment added but when I control the PR, there is no comment added by PrReview

**Reply by yehonatanblumenfeld** [reply]
Hey @Ahmet122, I also experience the same problem, i hope it will be fixed

**Comment by yehonatanblumenfeld** [comment]
Hey @Ahmet122 ,  i finally succeeded to solve this problem.

it was a problem with the permissions inside the yaml pipeline,
When you allow the 0auth Token in the classic pipeline, it gives the right permissions to the build service in order to add comments to the pr ,
However, when you use the yaml method. you must give the "Allow contribute to pull requests" the the specific build service,
so the name should be: Project Collection Build Service (name of organization).
you can see the the screenshot below the name of the service in my azure devops:

![Screenshot 2023-08-22 at 15 36 03](https://github.com/mlarhrouch/azure-pipeline-gpt-pr-review/assets/76685805/33d2e71c-18d8-4bfb-a1ee-18b5a7b4db35)

If this didn't help you, try searching the name Project Collection Build Service and you will see couple of results, just give them one by one until it works.

Another way to find out the specific build thats need the permission is to try to send a request using curl:
1. Send a request using curl from the **yaml pipeline** to a unique webhook like this website -> https://webhook.site/#!/d785ad98-9f04-4c93-a133-4e674905af5a,
2. In the webhook site, Extract the access token from the payload that you received.
3. Try to add a comment to the PR using this url with Postman and the access token of the response from last step -> 
'your-organization-uri/project-name/_apis/git/repositories/reponame/pullRequests/pr-id/threads?api-version=5.1'

- with this body:

comments: [
      {
        parentCommentId: 0,
        content: comment,
        commentType: 1
      }
    ],
    status: 1,
    threadContext: {
      filePath: fileName,
    }
  }
  
 - set the Autorization to bearer token with the access token you extracted.
 -  set the Content-Type to application/json

4. now send the request, if its worked, there is no problem with the permission, if not, you will get a message like "the specific object id doesn't have permissions"
5. copy the object id and paste it in the security page of the repository at the search bar like the picture below
![Screenshot 2023-08-22 at 15 54 53](https://github.com/mlarhrouch/azure-pipeline-gpt-pr-review/assets/76685805/612afcd7-d32a-4893-be95-be71dc437d27)

5. Press on the service and allow the Contribute to pull requests.

that should be it.
I hope it helps!

**Reply by yehonatanblumenfeld** [reply]
@mlarhrouch If you could please add logging to the project that will indicate if requests failed for some reason, cuz i needed to invest alot of time in the debugging process just to understand what was the problem. thanks!

**Reply by Ahmet122** [reply]
@yehonatanblumenfeld thank you, I will try your solution and write you back.

**Reply by Ahmet122** [reply]
Hi! I tried what you recommend but it didn't changes anything.
